tools/clang/lib/SPIRV/SpirvEmitter.cpp

```diff
@@ -11519,23 +11592,40 @@ SpirvEmitter::processIntrinsicAsType(const CallExpr *callExpr) {
   }
   case 3: {
     // Handling Method 6.
-    auto *value = doExpr(arg0);
-    auto *lowbits = doExpr(callExpr->getArg(1));
-    auto *highbits = doExpr(callExpr->getArg(2));
-    const auto uintType = astContext.UnsignedIntTy;
-    const auto uintVec2Type = astContext.getExtVectorType(uintType, 2);
-    auto *vecResult = spvBuilder.createUnaryOp(spv::Op::OpBitcast, uintVec2Type,
-                                               value, loc, range);
-    spvBuilder.createStore(
-        lowbits,
-        spvBuilder.createCompositeExtract(uintType, vecResult, {0},
-                                          arg0->getLocStart(), range),
-        loc, range);
-    spvBuilder.createStore(
-        highbits,
-        spvBuilder.createCompositeExtract(uintType, vecResult, {1},
-                                          arg0->getLocStart(), range),
-        loc, range);
+    const Expr *arg1 = callExpr->getArg(1);
+    SourceLocation arg1loc = arg1->getExprLoc();
+    SourceRange arg1range = arg1->getSourceRange();
+
+    const Expr *arg2 = callExpr->getArg(2);
+    SourceLocation arg2loc = arg2->getExprLoc();
+    SourceRange arg2range = arg2->getSourceRange();
+
+    SpirvInstruction *value = doExpr(arg0);
+    SpirvInstruction *lowbits = doExpr(arg1);
+    SpirvInstruction *highbits = doExpr(arg2);
+
+    QualType elemType = QualType();
+    uint32_t rowCount = 0;
+    uint32_t colCount = 0;
+
+    SpirvInstruction *lowbitsResult = nullptr;
+    SpirvInstruction *highbitsResult = nullptr;
+
+    if (isScalarType(argType)) {
+      splitDouble(value, lowbitsResult, highbitsResult, loc, range);
+    } else if (isVectorType(argType, &elemType, &rowCount)) {
+      splitDoubleVector(elemType, rowCount, arg1->getType(), value,
+                        lowbitsResult, highbitsResult, loc, range);
+    } else if (isMxNMatrix(argType, &elemType, &rowCount, &colCount)) {
+      splitDoubleMatrix(elemType, rowCount, colCount, arg1->getType(), value,
+                        lowbitsResult, highbitsResult, loc, range);
+    }
+
+    spvBuilder.createStore(lowbits, lowbitsResult, arg1loc, arg1range);
+    spvBuilder.createStore(highbits, highbitsResult, arg2loc, arg2range);
```

Comment:
For `asuint(d, x, y)`, `arg1loc` is the location of the `x` argument, and `arg2loc` is the location of `y`. I'm using them because I was thinking of the `OpStore` for each argument as being part of the code to handle that argument, though come to think of it, that's not how we normally treat `out` parameters. I'll replace these with `loc` and `range`.